src/apify/storages/dataset.py

```diff
@@ -1,19 +1,78 @@
 import csv
 import io
-from typing import AsyncIterator, Dict, List, Optional, Union
+import math
+from typing import AsyncIterator, Dict, Iterable, List, Optional, Union
 
 from apify_client import ApifyClientAsync
 from apify_client._utils import ListPage
 from apify_client.clients import DatasetClientAsync
 
 from .._types import JSONSerializable
-from .._utils import _wrap_internal
+from .._utils import _json_dumps, _wrap_internal
 from ..config import Configuration
+from ..consts import MAX_PAYLOAD_SIZE_BYTES
 from ..memory_storage import MemoryStorage
 from ..memory_storage.resource_clients import DatasetClient
 from .key_value_store import KeyValueStore
 from .storage_manager import StorageManager
 
+SAFETY_BUFFER_PERCENT = 0.01 / 100  # 0.01%
+EFFECTIVE_LIMIT_BYTES = MAX_PAYLOAD_SIZE_BYTES - math.ceil(MAX_PAYLOAD_SIZE_BYTES * SAFETY_BUFFER_PERCENT)
+
+
+def _check_and_serialize(item: JSONSerializable, index: Optional[int] = None) -> str:
+    """Accept a JSON serializable object as an input, validate its serializability and its serialized size against `EFFECTIVE_LIMIT_BYTES`."""
+    s = ' ' if index is None else f' at index {index} '
+
+    try:
+        payload = _json_dumps(item)
+    except Exception as e:
+        raise ValueError(f'Data item{s}is not serializable to JSON.\nCause: {e}')
+
+    length_bytes = len(payload.encode('utf-8'))
+    if length_bytes > EFFECTIVE_LIMIT_BYTES:
+        raise ValueError(f'Data item{s}is too large (size: {length_bytes} bytes, limit: {EFFECTIVE_LIMIT_BYTES} bytes)')
+
+    return payload
+
+
+def _chunk_by_size(items: List[str]) -> Iterable[str]:
```

Comment:
This is cool! I'm just a bit afraid about the memory usage, it means we'll have to keep all the stringified chunks in memory until we send them one by one, which could take quite a lot of memory.

Maybe it would be better to rewrite this to create an iterator, which would yield a chunk when it couldn't fit in any more items, and then forget that chunk?

Then in the future, we could much easier allow `Actor.push_data` to recieve an iterator of items, and it would chunk them and send them.

Reply:
Reworked it to use `yield`